### ISSUE TITLE
Define and implement the better location to download setuptools' eggs 

### ISSUE THREAD

**Issue by beraldoleal** [avocado-framework/avocado#5148]
**Is your feature request related to a problem? Please describe.**

As part of the #4934 , we need to download .eggs files from both: Avocado and Setuptools  in order to properly deploy Avocado inside an isolated environment. With #4692 merged, we are able to publish on GH releases our .eggs, and this will solve part of the problem. However we need the setuptools' eggs (our only dependency) as well.

**Describe the solution you'd like**

We could create targets to build .eggs for setuptools and publish on the same location, or...

**Describe alternatives you've considered**
Maybe setuptools maintainer has interest in having those targets there. Probably trickier, however It would free us from this burden.

**Comment by ana** [comment]
I went ahead and implemented the easy and fast solution, there is a fork of setuptools that we can sync regularly and build eggs with a github action https://github.com/avocado-framework/setuptools
See latest release https://github.com/avocado-framework/setuptools/releases/tag/v59.2.0

I don't know if this can work as a long term solution but I think it could help with this issue not blocking.

I agree the best solution would be at upstream but I don't see any reason why having this would be interesting for upstream. Their standard is wheels (https://packaging.python.org/discussions/wheel-vs-egg/)  and they already distribute them in pypi.

**Comment by beraldoleal** [comment]
> I went ahead and implemented the easy and fast solution, there is a fork of setuptools that we can sync regularly and build eggs with a github action https://github.com/avocado-framework/setuptools See latest release https://github.com/avocado-framework/setuptools/releases/tag/v59.2.0
> 
> I don't know if this can work as a long term solution but I think it could help with this issue not blocking.

I liked that. IMO, this will help as a short term solution, thank you @ana.

> I agree the best solution would be at upstream but I don't see any reason why having this would be interesting for upstream. Their standard is wheels (https://packaging.python.org/discussions/wheel-vs-egg/) and they already distribute them in pypi.

Even with wheels being a standard nowadays and a superior format for distribution, afaict, eggs had one plus feature: it is also a runtime installation format, it was designed to be importable. We can run directly from eggs, without the need from installing it. IIRC, that was the reason we choose eggs over wheels. So this *could be* one reason for them releasing the eggs as well. However, trying pushing this upstream just for us, could "break some eggs":

From PEP427:

> Technically, due to the combination of supporting installation via simple extraction and using an archive format that is compatible with zipimport, a subset of wheel files do support being placed directly on sys.path. However, while this behaviour is a natural consequence of the format design, actually relying on it is generally discouraged.

This might be true even for eggs, not sure. So unless I'm missing something, I would say your suggestion it will work for us, until we came up with a better distribution model that maybe we don't need running the runners directly from the .zip.

Let's give some time to see if others comment here, otherwise I think we can close this issue.

**Comment by ana** [comment]
Btw, we can remove the `.post20211119` from the egg names or choose to have something else.
It's defined at https://github.com/avocado-framework/setuptools/blob/main/setup.cfg#L138

**Comment by ana** [comment]
> 
> Even with wheels being a standard nowadays and a superior format for distribution, afaict, eggs had one plus feature: it is also a runtime installation format, it was designed to be importable. We can run directly from eggs, without the need from installing it. IIRC, that was the reason we choose eggs over wheels. So this _could be_ one reason for them releasing the eggs as well. However, trying pushing this upstream just for us, could "break some eggs":
> 

Yeah, I think the fact is the only format that allows distribution and runtime is one of the reasons why eggs are still around.
Without a deep knowledge of how things are these days in the python development community, I'd feel like walking on eggshells asking them about this.

**Comment by ana** [comment]
@beraldoleal I have updated the github action to override the `egg_info` options from `setup.cfg`. There isn't a date anymore, see: https://github.com/avocado-framework/setuptools/releases/tag/v59.2.0 


@clebergnu As we discussed yesterday, if you have something to say about this, please share it now :D

**Comment by beraldoleal** [comment]
Thanks @ana.

**Comment by ana** [comment]
This solution might evolve in the future, but for now we can consider this done.
